src/core/compact_object.cc

```diff
@@ -138,39 +138,74 @@ inline void FreeObjZset(unsigned encoding, void* ptr) {
   }
 }
 
+pair<void*, bool> DefragStrMap2(StringMap* sm, float ratio) {
+  bool realloced = false;
+
+  for (auto it = sm->begin(); it != sm->end(); ++it)
+    realloced |= it.ReallocIfNeeded(ratio);
+
+  return {sm, realloced};
+}
+
+pair<void*, bool> DefragListPack(uint8_t* lp, float ratio) {
+  if (!zmalloc_page_is_underutilized(lp, ratio))
+    return {lp, false};
+
+  size_t lp_bytes = lpBytes(lp);
+  uint8_t* replacement = lpNew(lpBytes(lp));
+  memcpy(replacement, lp, lp_bytes);
+  lpFree(lp);
+
+  return {replacement, true};
+}
+
+pair<void*, bool> DefragIntSet(intset* is, float ratio) {
+  if (!zmalloc_page_is_underutilized(is, ratio))
+    return {is, false};
+
+  const size_t blob_len = intsetBlobLen(is);
+  intset* replacement = (intset*)zmalloc(blob_len);
```

Comment:
we are allocating here another set with same size, but what if this set is very big? it can cause spike in rss.
for listpack its ok because we use it for small hsets

Reply:
Imo I thought the exact same thing but we do the same for other DefragLogic. Maybe we can do this as an improvement for a future PR? That is, do not Defrag relateively big data structures ?

Reply:
what other defrag logic will ? as I wrote for listpack it is not risk because it is limited in size,
ReallocIfNeeded in StrMap is itereating on all elements so we are not allocating twice the size of the entire map right? regarding string type reallocation I dont think thats an issue because I dont think uses use huge strings thats probably not a real use case

Reply:
oh inset is also limited with number of elements that in this case there is no issue